### PR TITLE
Feat/cdn

### PR DESCRIPTION
**Pull request by rmocanu-ionos**
## Checklist

- [ ] PR name added as appropriate (e.g. `feat:`/`fix:`/`doc:`/`test:`/`refactor:`)
- [ ] Tests added or updated
- [ ] Documentation updated
- [ ] Changelog updated and version incremented (label: upcoming release)
- [ ] Github Issue linked if any
- [ ] Jira task updated

### CODE REVIEW

File: ionoscloud/data_source_distribution.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+							!partialMatch && strings.EqualFold(*distributionItem.Properties.Domain, domain)) {
+						distributionsByDomain = append(distributionsByDomain, distributionItem)
+					}
+				}
+				if distributionsByDomain != nil && len(distributionsByDomain) > 0 {
+					results = distributionsByDomain
+				} else {
+					return diag.FromErr(fmt.Errorf("no distribution found with the specified criteria: domain = %v", domain))
+				}
+			}
```
Comment (avirtopeanu-ionos) [inline ionoscloud/data_source_distribution.go:172]: (just a nitpick) these nested conditions are a bit hard to read, as well as being deeply nested and in a large function

```
if distributions.Items != nil && len(*distributions.Items) > 0 {
	var distributionsByDomain []cdn.Distribution
	for _, item := range *distributions.Items {
		if isValidDomain(item.Properties, domain, partialMatch) {
			distributionsByDomain = append(distributionsByDomain, item)
		}
	}
	if len(distributionsByDomain) > 0 {
		results = distributionsByDomain
	} else {
		return diag.FromErr(fmt.Errorf("no distribution found with the specified criteria: domain = %v", domain))
	}
}

func isValidDomain(properties *cdn.Properties, domain string, partialMatch bool) bool {
	if properties == nil || properties.Domain == nil {
		return false
	}
	if partialMatch {
		return strings.Contains(*properties.Domain, domain)
	}
	return strings.EqualFold(*properties.Domain, domain)
}
```

**Review by avirtopeanu-ionos — APPROVED** [review]

File: ionoscloud/data_source_cdn_distribution.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+		var results []cdn.Distribution
+
+		distributions, _, err := client.SdkClient.DistributionsApi.DistributionsGet(ctx).Execute()
+		if err != nil {
+			diags := diag.FromErr(fmt.Errorf("an error occurred while fetching container distributions: %w", err))
+			return diags
+		}
+
+		results = *distributions.Items
+		if domainOk {
```
Comment (digna-ionos) [inline ionoscloud/data_source_cdn_distribution.go:154]: Do we need this check?

File: ionoscloud/data_source_cdn_distribution.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+			if distributions.Items != nil && len(*distributions.Items) > 0 {
+				var distributionsByDomain []cdn.Distribution
+
+				for _, distributionItem := range *distributions.Items {
+					if isValidDomain(distributionItem.Properties, domain, partialMatch) {
+						distributionsByDomain = append(distributionsByDomain, distributionItem)
+					}
+				}
+				if len(distributionsByDomain) > 0 {
```
Comment (digna-ionos) [inline ionoscloud/data_source_cdn_distribution.go:167]: here we can return early on len(distributionsByDomain) == 0 and we can avoid the nested else.

File: ionoscloud/resource_cdn_distribution.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+			Domain: &distributionDomain,
+		},
+	}
+
+	if attr, ok := d.GetOk("certificate_id"); ok {
+		attrStr := attr.(string)
+		distribution.Properties.CertificateId = &attrStr
+	}
+
+	if routingRules, err := cdnService.GetRoutingRulesData(d); err == nil {
```
Comment (digna-ionos) [inline ionoscloud/resource_cdn_distribution.go:136]: better check for err != nil

File: ionoscloud/resource_cdn_distribution.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+			Domain: &distributionDomain,
+		},
+	}
+
+	if attr, ok := d.GetOk("certificate_id"); ok {
+		attrStr := attr.(string)
+		request.Properties.CertificateId = &attrStr
+	}
+
+	if routingRules, err := cdnService.GetRoutingRulesData(d); err == nil {
```
Comment (digna-ionos) [inline ionoscloud/resource_cdn_distribution.go:198]: same here

File: ionoscloud/data_source_cdn_distribution.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	cdn "github.com/ionos-cloud/sdk-go-cdn"
+
+	"github.com/ionos-cloud/terraform-provider-ionoscloud/v6/services"
+	cdnService "github.com/ionos-cloud/terraform-provider-ionoscloud/v6/services/cdn"
+)
+
+func dataSourceCdnDistribution() *schema.Resource {
```
[the remark this answers is not in the document]
Comment (rmocanu-ionos) [reply ionoscloud/data_source_cdn_distribution.go:18]: Done

**Review by digna-ionos — APPROVED** [review]

File: ionoscloud/resource_cdn_distribution.go (hunk 4 above)
Comment (rmocanu-ionos) [reply ionoscloud/resource_cdn_distribution.go]: Done

File: ionoscloud/resource_cdn_distribution.go (hunk 5 above)
Comment (rmocanu-ionos) [reply ionoscloud/resource_cdn_distribution.go]: Done

File: ionoscloud/data_source_cdn_distribution.go (hunk 2 above)
Comment (rmocanu-ionos) [reply ionoscloud/data_source_cdn_distribution.go:154]: I think changing this should be done when we can change all data sources, as they are all similar

File: ionoscloud/data_source_cdn_distribution.go (hunk 3 above)
Comment (rmocanu-ionos) [reply ionoscloud/data_source_cdn_distribution.go:167]: I think changing this should be done when we can change all data sources, as they are all similar

File: ionoscloud/resource_cdn_distribution.go (hunk 5 above)
Comment (rmocanu-ionos) [reply ionoscloud/resource_cdn_distribution.go]: Changing this required creating a new variable outside the the if so I decided to undo it

**Review by digna-ionos — DISMISSED** [review]

**Pull request merged**